xds/src/test/java/io/grpc/xds/FilterChainMatchingProtocolNegotiatorsTest.java

```diff
@@ -0,0 +1,1318 @@
+import io.grpc.netty.InternalProtocolNegotiationEvent;
+import io.grpc.netty.InternalProtocolNegotiator.ProtocolNegotiator;
+import io.grpc.netty.ProtocolNegotiationEvent;
+import io.grpc.xds.EnvoyServerProtoData.CidrRange;
+import io.grpc.xds.EnvoyServerProtoData.DownstreamTlsContext;
+import io.grpc.xds.EnvoyServerProtoData.FilterChain;
+import io.grpc.xds.EnvoyServerProtoData.FilterChainMatch;
+import io.grpc.xds.Filter.NamedFilterConfig;
+import io.grpc.xds.FilterChainMatchingProtocolNegotiators.FilterChainMatchingHandler;
+import io.grpc.xds.FilterChainMatchingProtocolNegotiators.FilterChainMatchingHandler.FilterChainSelector;
+import io.grpc.xds.internal.sds.CommonTlsContextTestsUtil;
+import io.grpc.xds.internal.sds.SslContextProviderSupplier;
+import io.netty.channel.ChannelHandler;
+import io.netty.channel.ChannelHandlerContext;
+import io.netty.channel.ChannelInboundHandlerAdapter;
+import io.netty.channel.ChannelPipeline;
+import io.netty.channel.ChannelPromise;
+import io.netty.channel.embedded.EmbeddedChannel;
+import io.netty.handler.codec.http2.DefaultHttp2Connection;
+import io.netty.handler.codec.http2.DefaultHttp2ConnectionDecoder;
+import io.netty.handler.codec.http2.DefaultHttp2ConnectionEncoder;
+import io.netty.handler.codec.http2.DefaultHttp2FrameReader;
+import io.netty.handler.codec.http2.DefaultHttp2FrameWriter;
+import io.netty.handler.codec.http2.Http2ConnectionDecoder;
+import io.netty.handler.codec.http2.Http2ConnectionEncoder;
+import io.netty.handler.codec.http2.Http2Settings;
+import java.net.InetSocketAddress;
+import java.net.SocketAddress;
+import java.net.UnknownHostException;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collections;
+import org.junit.Rule;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.junit.runners.JUnit4;
+import org.mockito.Mock;
+import org.mockito.junit.MockitoJUnit;
+import org.mockito.junit.MockitoRule;
+
+@RunWith(JUnit4.class)
+public class FilterChainMatchingProtocolNegotiatorsTest {
+  @Rule
+  public final MockitoRule mocks = MockitoJUnit.rule();
+
+  private final GrpcHttp2ConnectionHandler grpcHandler =
+          FakeGrpcHttp2ConnectionHandler.newHandler();
+  @Mock private TlsContextManager tlsContextManager;
+  private ProtocolNegotiationEvent event = InternalProtocolNegotiationEvent.getDefault();
+  private ChannelPipeline pipeline;
+  private EmbeddedChannel channel;
+  private ChannelHandlerContext channelHandlerCtx;
+  @Mock
+  private ProtocolNegotiator mockDelegate;
+  private final SettableFuture<SslContextProviderSupplier> sslSet = SettableFuture.create();
+  private static final HttpConnectionManager HTTP_CONNECTION_MANAGER = createRds("routing-config");
+  private static final String LOCAL_IP = "10.1.2.3";  // dest
+  private static final String REMOTE_IP = "10.4.2.3"; // source
+  private static final int PORT = 7000;
+
+  @Test
+  public void filterChainMatch() throws Exception {
```

Comment:
Most tests here are basically testing just `FilterChainSelector.select()`. 

Have you considered doing just one test to test the whole flow of `FilterChainMatchingHandler` using a `mockDelegate` and the other tests just testing `FilterChainSelector.select()`? All you need for those tests is just a `FilterChainSelector` with test filterChains and local and  remote IP address values.

We don't gain anything by adding  extra test code for the mock `ChannelHandler`, mock Channel, pipeline etc.

Reply:
I am not insisting and I agree "it does not hurt" but  if there is some change in the `FilterChainMatchingHandler` logic in how it calls the FilterChainSelector there is the overhead of changing all the tests. Also for future simplicity if you want to add more functionality and tests having a simpler test setup might make your own job easier